opensrp-fp/src/main/java/org/smartregister/fp/common/task/FinalizeVisitFormTask.java

```diff
@@ -0,0 +1,174 @@
+        this.baseEntityId = baseEntityId;
+        this.contactNo = contactNo;
+        this.contact = contact;
+        this.jsonCurrentState = jsonCurrentState;
+        fpFormUtils = new FPFormUtils();
+
+        progressDialog = new ProgressDialog(context);
+        progressDialog.setCancelable(false);
+        progressDialog.setTitle(context.getString(com.vijay.jsonwizard.R.string.loading));
+        progressDialog.setMessage(context.getString(R.string.finalizing_form_message));
+        progressDialog.show();
+    }
+
+    @Override
+    protected Void doInBackground(Void... voids) {
+
+        contact.setJsonForm(fpFormUtils.addFormDetails(jsonCurrentState));
+        contact.setContactNumber(contactNo);
+        FPFormUtils.persistPartial(baseEntityId, contact);
+
+        /*Map<String, String> clientProfileDetail = PatientRepository.getClientProfileDetails(baseEntityId);
+        ContactInteractor contactInteractor = new ContactInteractor();
+        contactInteractor.finalizeContactForm(clientProfileDetail, this);*/
+
+        clientProfileDetail = PatientRepository.getClientProfileDetails(baseEntityId);
+
+        int gestationAge = getGestationAge(clientProfileDetail);
+        boolean isFirst = TextUtils.equals("1", clientProfileDetail.get(DBConstantsUtils.KeyUtils.NEXT_CONTACT));
+        ContactRule contactRule = new ContactRule(gestationAge, isFirst, baseEntityId);
+
+        List<Integer> integerList = FPLibrary.getInstance().getFPRulesEngineHelper()
+                .getContactVisitSchedule(contactRule, ConstantsUtils.RulesFileUtils.CONTACT_RULES);
+
+        int nextContactVisitWeeks = integerList.get(0);
+
+        LocalDate localDate = new LocalDate(clientProfileDetail.get(DBConstantsUtils.KeyUtils.EDD));
+        String nextContactVisitDate = localDate.minusWeeks(ConstantsUtils.DELIVERY_DATE_WEEKS).plusWeeks(nextContactVisitWeeks).toString();
+        int nextContact = getNextContact(clientProfileDetail);
+        WomanDetail womanDetail = getWomanDetail(baseEntityId, nextContactVisitDate, nextContact);
+
+        PatientRepository.updateContactVisitDetails(womanDetail, true);
+
+        try {
+            processFormFieldKeyValues(baseEntityId, new JSONObject(contact.getJsonForm()), String.valueOf(contactNo));
+        }
+        catch (Exception ex) {
+            Timber.e(ex);
+        }
+
+        return null;
+    }
+
+    @Override
+    protected void onPostExecute(Void aVoid) {
+        super.onPostExecute(aVoid);
+        if (progressDialog != null && progressDialog.isShowing()) {
+            progressDialog.dismiss();
+        }
+        Utils.navigateToProfile(context, clientProfileDetail);
+    }
+
+
+    private int getNextContact(Map<String, String> details) {
+        int nextContact = details.containsKey(DBConstantsUtils.KeyUtils.NEXT_CONTACT) && details.get(DBConstantsUtils.KeyUtils.NEXT_CONTACT) != null ? Integer.valueOf(details.get(DBConstantsUtils.KeyUtils.NEXT_CONTACT)) : 1;
```

Comment:
Possible `NumberFormatException`
Get containsKey + get + get results to to 0(n) + 0(n) +0(n) time to access the variable

See this sample. those three operations result to 0(n)

```
    private int getNextContact1(Map<String, String> details) {
        String nextContactString = details.get(DBConstantsUtils.KeyUtils.NEXT_CONTACT);
        if(nextContactString != null){
            try{
                Integer x = Integer.valueOf(nextContactString);
                return  ++x;
            }catch(NumberFormatException ex){
                Timber.e(ex);
            }
        }
        return 1;
    }
```

Reply:
Fixed that.